src/App/Pages/Settings/ExportVaultPageViewModel.cs

```diff
@@ -0,0 +1,144 @@
+            _cryptoService = ServiceContainer.Resolve<ICryptoService>("cryptoService");
+            _exportService = ServiceContainer.Resolve<IExportService>("exportService");
+
+            PageTitle = AppResources.ExportVault;
+            TogglePasswordCommand = new Command(TogglePassword);
+            ExportVaultCommand = new Command(async () => await ExportVaultAsync());
+
+            FileFormatOptions = new List<KeyValuePair<string, string>>
+            {
+                new KeyValuePair<string, string>("json", ".json"),
+                new KeyValuePair<string, string>("csv", ".csv")
+            };
+        }
+
+        public async Task InitAsync()
+        {
+            FileFormatSelectedIndex = FileFormatOptions.FindIndex(k => k.Key == "json");
+        }
+
+        public List<KeyValuePair<string, string>> FileFormatOptions { get; set; }
+
+        public int FileFormatSelectedIndex
+        {
+            get => _fileFormatSelectedIndex;
+            set { SetProperty(ref _fileFormatSelectedIndex, value); }
+        }
+
+        public bool ShowPassword
+        {
+            get => _showPassword;
+            set => SetProperty(ref _showPassword, value,
+                additionalPropertyNames: new string[] {nameof(ShowPasswordIcon)});
+        }
+
+        public string MasterPassword
+        {
+            get => _masterPassword;
+            set => SetProperty(ref _masterPassword, value);
+        }
+
+        public Command TogglePasswordCommand { get; }
+
+        public string ShowPasswordIcon => ShowPassword ? "" : "";
+
+        public void TogglePassword()
+        {
+            ShowPassword = !ShowPassword;
+            (Page as ExportVaultPage).MasterPasswordEntry.Focus();
+        }
+
+        public Command ExportVaultCommand { get; }
+
+        public async Task ExportVaultAsync()
+        {
+            if(string.IsNullOrEmpty(_masterPassword))
+            {
+                _platformUtilsService.ShowToast("error", _i18nService.T("AnErrorHasOccurred"),
+                    _i18nService.T("InvalidMasterPassword"));
+                return;
+            }
+
+            var keyHash = await _cryptoService.HashPasswordAsync(_masterPassword, null);
+            MasterPassword = string.Empty;
+            
+            var storedKeyHash = await _cryptoService.GetKeyHashAsync();
+            if(storedKeyHash != null && keyHash != null && storedKeyHash == keyHash)
+            {
+                try
+                {
+                    var data = _exportService.GetExport(FileFormatOptions[FileFormatSelectedIndex].Key);
+                    var fileFormat = FileFormatOptions[FileFormatSelectedIndex].Key;
+                    _defaultFilename = _exportService.GetFileName(null, fileFormat);
+                    _exportResult = Encoding.ASCII.GetBytes(data.Result);
+
+                    if(!_deviceActionService.SaveFile(_exportResult, null, _defaultFilename, null))
```

Comment:
I am a little confused about why we have `_deviceActionService.SaveFile` here and in the `SaveFileSelected` method below. are there two paths for saving a file? I see the code-behind file calls into `SaveFileSelected` as well when a message is received.

Reply:
TLDR: It's a continuation of the save flow, specific to Android.

`_deviceActionService.SaveFile` starts the process.  For iOS it simply follows the existing `_deviceActionService.OpenFile` flow since that gives us what we want (options for saving locally as well as cloud services).  On Android however new logic was required to get the same options (via storage access framework).  Once the user selects a save destination, the uri & filename is returned via `MainActivity.OnActivityResult` and forwarded to `ExportVaultPageViewModel.SaveFileSelected` via `_messagingService.Send("selectSaveFileResult")`.